src/object-store/src/util/thread_id.hpp

```diff
@@ -0,0 +1,104 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_THREAD_ID_HPP
+#define REALM_THREAD_ID_HPP
+
+#include <realm/util/features.h>
+
+#include <atomic>
+
+#if __has_feature(tls) || __has_feature(cxx_thread_local)
+#define REALM_HAS_THREAD_LOCAL 1
+#else
+#define REALM_HAS_THREAD_LOCAL 0
+#include <pthread.h>
+#endif
+
+namespace realm {
+
+using thread_id_t = std::size_t;
+
+namespace _impl {
```

Comment:
I actually disagree because `get_thread_id()` is a very hot function that should be inlined, and have its implementation inlined, as much as possible. Also, `ThreadLocal` is a class template.